### PR TITLE
Extend Slideshow block with source option

### PR DESCRIPTION
**Pull request by madalingorbanescu**

### CODE REVIEW

File: packages/block-library/src/blocks/slideshow/preview.js [hunk 1]
```diff
@@ -104,33 +106,31 @@ const SlideshowPreview = class extends Component {
...
+          <a> {i + 1}</a>
+        );
+      },
+		  infinite: true,
+      arrows: true,
+      dots: true,
+      speed: 0
+    };
+
+    let slide = getGallerySlideData( galleryImages, this.props );
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/preview.js:121]:
```js
let slides = getSlidesFromGalleryImages( ...
```

File: packages/block-library/src/blocks/slideshow/edit.js [hunk 2]
```diff
@@ -88,8 +74,6 @@ class Edit extends Component {
 				<SlideshowPreview
 					{ ...newProps }
 					previewImage={ galleryImages[ selectedIndex ] }
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/edit.js:76]: selectedIndex seems to still be relevant for editing the focal point for each image, you could try to refactor this component into a functional component and use `useState()` for cleaner code

Comment (razwan) [reply packages/block-library/src/blocks/slideshow/edit.js:76]: setIndex function passed to the GalleryPreview component inside InspectorControls should also trigger slick to go to that specific slide

File: packages/block-library/src/blocks/slideshow/frontend.js [hunk 3]
```diff
@@ -70,7 +74,7 @@ const TRANSITION_EASING = "easeInOutCirc";
 		var slideMaxHeight = 0;
 		var maxAspectRatio = 0;
 
-		$block.find( SLIDE_SELECTOR ).each( function( i, obj ) {
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/frontend.js:73]: each should iterate over `SLIDE_SELECTOR` and not `SLIDER_SELECTOR`

File: packages/block-library/src/blocks/slideshow/init.php [hunk 4]
```diff
@@ -84,69 +93,27 @@ class="<?php echo esc_attr( join( ' ', $classes ) ); ?>"
...
-                                    <?php
-                                    novablocks_the_media_title( $media, '<h2>', '</h2>' );
-                                    novablocks_the_media_caption( $media );
-                                    ?>
-                                </div>
-	                        </div>
-                        </div>
-                    </div>
-				<?php } ?>
+
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/init.php:96]: There should be a check for when there are no slides and not output and / or initialize the slider in this case.

File: packages/block-library/src/blocks/slideshow/edit.js [hunk 5]
```diff
@@ -88,8 +74,6 @@ class Edit extends Component {
...
-					onPrevArrowClick={ this.onPrevArrowClick.bind( this ) }
-					onNextArrowClick={ this.onNextArrowClick.bind( this ) }
 				/>
 
 				<InspectorControls { ...{ ...newProps, setIndex, selectedIndex } } />
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/edit.js:79]: Change Media button should not be displayed inside block controls when the source attribute is set to posts

File: packages/block-library/src/blocks/slideshow/init.php [hunk 6]
```diff
@@ -158,3 +125,112 @@ class="<?php echo esc_attr( join( ' ', $classes ) ); ?>"
 		return ob_get_clean();
 	}
 }
+
+function novablocks_get_post_slide_markup( $post, $attributes ) {
+
+	$slider_background_src = get_the_post_thumbnail_url( $post );
+	$thumbnail = wp_get_attachment_metadata( get_post_thumbnail_id($post) );
+
+	$width = $thumbnail['sizes']['large']['width'];
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/init.php:134]: There is the change for the attachment not having the large image size.

Markup output should be abstracted into a function that receives $image or $image_id and $content and outputs them with the proper wrappers:
```php
	<div class="<?php echo esc_attr( $className ); ?>">
		<div class="novablocks-slideshow__slide-wrap">
			<div class="novablocks-slideshow__background novablocks-u-background">
				<div class="novablocks-mask">
					<div class="novablocks-parallax">
					    <?php novablocks_output_slideshow_slide_background( $image or $image_id ); ?>
					....
			<div class="novablocks-slideshow__foreground novablocks-foreground novablocks-u-content-padding novablocks-u-content-align">
				<div class="novablocks-slideshow__inner-container novablocks-u-content-width">
				<?php echo $content; ?>
				...
```
in order to not repeat yourself and have consistent markup for both use cases

File: packages/block-library/src/blocks/slideshow/inspector-controls.js [hunk 7]
```diff
@@ -105,27 +122,94 @@ const SlideshowInspectorControls = function( props ) {
...
+            </ControlsGroup>
+            <MetaSource { ...props } />
+          </ControlsTab>
+        </ControlsSection>
+      }
 		</Fragment>
 	);
 };
 
+const MetaSource = ( props ) => {
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/inspector-controls.js:171]: MetaSource and ToggleGroup should be moved inside a hook in the block-editor package. the ToggleGroup component should receive a function through props in order to filter the displayed toggles. At the same time, attributes needed for these components should be moved to that hook as well.

File: packages/block-library/src/blocks/slideshow/preview.js [hunk 8]
```diff
@@ -104,33 +106,31 @@ const SlideshowPreview = class extends Component {
...
-						</div>
-					</div>
-					{ galleryImages.length > 1 && <div className="novablocks-slideshow__controls">
-						<div className="novablocks-slideshow__arrow novablocks-slideshow__arrow--prev novablocks-slideshow__arrow--disabled" onClick={ this.props.onPrevArrowClick }></div>
-						<div className="novablocks-slideshow__arrow novablocks-slideshow__arrow--next novablocks-slideshow__arrow--disabled" onClick={ this.props.onNextArrowClick }></div>
-					</div> }
-				</div> }
-				{ ! galleryImages.length && <GalleryPlaceholder { ...this.props } /> }
+          <div className={ classes.join( ' ' ) } style={ styles.slideshow }>
+             <Slider className="novablocks-slideshow__slider" {...sliderSettings}>
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/preview.js:130]: You should start with this. For cleaner code you should split the two use cases into separate components like so.
```js
            <GallerySlider { ...props } />
            <PostsSlider { ...props } />
```

File: packages/block-library/src/blocks/slideshow/preview.js [hunk 9]
```diff
@@ -146,3 +146,23 @@ const SlideshowPreview = class extends Component {
 };
 
 export default SlideshowPreview;
+
+const getPostSlideData = ( posts, props ) => {
+  return (
+    posts.map( ( post, id ) => {
+      return (
+        <Slide id={id} className="novablocks-slideshow__slide" {...props} image={ wp.data.select( 'core').getMedia(post.featured_media) } post={ post } type="post"/>
```
Comment (razwan) [inline packages/block-library/src/blocks/slideshow/preview.js:154]: the image prop should be standardized before passing it to the Slide component and not inside it

Comment (razwan) [reply packages/block-library/src/blocks/slideshow/preview.js:154]: Also... the best case scenario is when the Slide component receives only two props `media` and `content`. You could try to reach this scenario

**Review by razwan — CHANGES_REQUESTED** [review]

**Comment by madalingorbanescu** [comment]
I am going to close this PR because those changes were moved on `try/supernova-block` branch.

**Pull request closed without merging**